File: Find_the_Missing_Number_in_a_sorted_array.py
def findMissingNumber(arr):
    low, high = 0, len(arr)-1
    
    ### Binary Search
    while low <= high:
        mid = low + (high-low)//2
        if arr[mid] == mid+1:
            low = mid+1
        else:
            high = mid-1
    return low+1

# Driver Code 
test_1 = [1, 2, 3, 4, 5, 6, 8] 
test_2 = [2, 3, 4, 5, 6, 7] 
test_3 = [1, 2, 3, 5, 6, 7] 
print("Missing number:", findMissingNumber(test_3)) 

### Complexity Analysis ###
# Time Complexity: O(logN) --> Binary Search
# Space Complexity: O(1)

File: Find the Missing Number in a sorted array.java
public class FindTheMissingSortedArrayAgain {

  public int findMissingElement(int[] arr)
  {
    if(arr==null || arr.length==0 || arr.length==1)
    {
      return -1;
    }
    int high = arr.length-1;
    int low = 0;
    int base = arr[0];
    int end = arr[arr.length-1];
    while(low<=high) {
      int mid = low + (high - low) / 2;
      if(arr[mid] == base+mid )
      {
        if(arr[mid+1] == arr[mid]+1)
        {
        // right side
        low = mid+1;
        }
        else {
          return arr[mid]+1;
        }
      }
      else {
        if(arr[mid] == arr[mid-1]+1)
        {
          high = mid-1;
        }
        else {
          return arr[mid]-1;
        }
      }
    }
    return -1;

  }
}

 class Solution
{
  public static void main(String[] args)
  {
    FindTheMissingSortedArrayAgain a = new FindTheMissingSortedArrayAgain();
    System.out.println(a.findMissingElement(new int[]{22,23,24,25,26,27,29}));

  }
}

